File: examples/node-express-demo/statlite-metrics.js
const { performance } = require("node:perf_hooks");

const METRICS_PATH = "/statlite/metrics";
const startedAt = new Date(Date.now() - process.uptime() * 1000);

const counters = {
  requestsTotal: 0,
  responses404Total: 0,
  responses4xxTotal: 0,
  responses5xxTotal: 0,
  requestDurationSecondsTotal: 0,
};

let previousCpu = process.cpuUsage();
let previousCpuTime = performance.now();

function record(statusCode, durationSeconds) {
  counters.requestsTotal += 1;
  counters.requestDurationSecondsTotal += durationSeconds;
  if (statusCode === 404) counters.responses404Total += 1;
  if (statusCode >= 400 && statusCode < 500) counters.responses4xxTotal += 1;
  if (statusCode >= 500 && statusCode < 600) counters.responses5xxTotal += 1;
}

function statliteMetricsMiddleware(req, res, next) {
  if (req.path === METRICS_PATH) return next();

  const requestStarted = performance.now();
  res.once("finish", () => {
    record(res.statusCode, (performance.now() - requestStarted) / 1000);
  });
  next();
}

function snapshot() {
  const now = performance.now();
  const cpu = process.cpuUsage();
  const elapsedSeconds = (now - previousCpuTime) / 1000;
  const cpuSeconds =
    (cpu.user - previousCpu.user + cpu.system - previousCpu.system) / 1e6;
  const processCpuUsage = elapsedSeconds > 0 ? cpuSeconds / elapsedSeconds : 0;
  previousCpu = cpu;
  previousCpuTime = now;

  return {
    schema: "statlite-metrics/v1",
    status: "UP",
    started_at: startedAt.toISOString(),
    metrics: {
      requests_total: counters.requestsTotal,
      responses_404_total: counters.responses404Total,
      responses_4xx_total: counters.responses4xxTotal,
      responses_5xx_total: counters.responses5xxTotal,
      request_duration_seconds_total: counters.requestDurationSecondsTotal,
      process_cpu_usage: processCpuUsage,
      runtime_heap_used_bytes: process.memoryUsage().heapUsed,
      uptime_seconds: process.uptime(),
    },
  };
}

function statliteMetricsEndpoint(req, res) {
  res.json(snapshot());
}

module.exports = {
  METRICS_PATH,
  statliteMetricsEndpoint,
  statliteMetricsMiddleware,
};

File: examples/python-django-demo/demo/statlite_metrics.py
import threading
import time
import tracemalloc
from datetime import datetime, timezone

from django.http import JsonResponse
from django.views.decorators.http import require_GET


METRICS_PATH = "/statlite/metrics"


class StatLiteMetrics:
    def __init__(self):
        self._lock = threading.Lock()
        self._started_at = datetime.now(timezone.utc)
        self._started_monotonic = time.monotonic()
        self._previous_cpu = time.process_time()
        self._previous_cpu_time = self._started_monotonic
        if not tracemalloc.is_tracing():
            tracemalloc.start()
        self._requests = 0
        self._responses_404 = 0
        self._responses_4xx = 0
        self._responses_5xx = 0
        self._duration_seconds = 0.0

    def record(self, status_code, duration_seconds):
        with self._lock:
            self._requests += 1
            self._duration_seconds += duration_seconds
            if status_code == 404:
                self._responses_404 += 1
            if 400 <= status_code < 500:
                self._responses_4xx += 1
            if 500 <= status_code < 600:
                self._responses_5xx += 1

    def snapshot(self):
        runtime_heap_used_bytes, _peak = tracemalloc.get_traced_memory()
        with self._lock:
            now = time.monotonic()
            cpu = time.process_time()
            elapsed = now - self._previous_cpu_time
            process_cpu_usage = (
                (cpu - self._previous_cpu) / elapsed if elapsed > 0 else 0.0
            )
            self._previous_cpu = cpu
            self._previous_cpu_time = now
            metrics = {
                "requests_total": self._requests,
                "responses_404_total": self._responses_404,
                "responses_4xx_total": self._responses_4xx,
                "responses_5xx_total": self._responses_5xx,
                "request_duration_seconds_total": self._duration_seconds,
                "process_cpu_usage": process_cpu_usage,
                "runtime_heap_used_bytes": runtime_heap_used_bytes,
                "uptime_seconds": max(0.0, now - self._started_monotonic),
            }

        return {
            "schema": "statlite-metrics/v1",
            "integration": "django",
            "status": "UP",
            "started_at": self._started_at.isoformat(timespec="microseconds").replace(
                "+00:00", "Z"
            ),
            "metrics": metrics,
        }


metrics = StatLiteMetrics()


class StatLiteMetricsMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        if request.path == METRICS_PATH:
            return self.get_response(request)

        started = time.monotonic()
        try:
            response = self.get_response(request)
        except Exception:
            metrics.record(500, time.monotonic() - started)
            raise

        metrics.record(response.status_code, time.monotonic() - started)
        return response


@require_GET
def statlite_metrics_view(request):
    return JsonResponse(metrics.snapshot())
